Technotes: Change management and Demand oriented schema design (#335)
apollographql/docs: src/content/technotes/TN0028-change-management.mdx

    C -.-> A
    B --> D(REST API)
    D -.-> B
    D --> E(Database)
    E -.-> D
    C --> F(gRPC API)
    F -.-> C
    F --> G(Third party API)
    G -.-> F
```

Allowing teams to independently update the components they own is a key benefit of federation. However, understanding and de-risking changes to a component can be challenging, especially because changes might require cascading changes in other components. This is especially true for GraphQL schema definitions, which must be coordinated between subgraph services and the Router.

This section covers the most common change management scenarios in a federated architecture using the Managed Federation features of Apollo GraphOS. We'll consider the proper sequencing of release events, how Apollo tooling makes managing changes easier, and when progressive delivery techniques like blue-green deploys are useful.

Specifically, we'll examine scenarios that:

- Involve a schema change
  - [Backward compatible subgraph schema changes](#backward-compatible-subgraph-schema-changes)
  - [Backward _incompatible_ subgraph schema changes](#backward-incompatible-subgraph-schema-changes)
- Involve schema changes in multiple subgraphs

>   - [Changes to shared types](#changes-to-shared-types)

Reviewer: might want to double-check this anchor, it's not rendering as a link for me

Reviewer: oh you removed the section so this bullet is no longer relevant

Author: good catch, forgot to remove this. will remove this and the other later sections

  - [Changes to query planning](#changes-to-query-planning-field-migrations-entity-ownership-changes-etc)
- Don't involve schema changes
  - [Changes to a data source](#changes-to-a-data-source)
  - [Changes to subgraph resolvers](#changes-to-subgraph-resolvers)
  - [Changes to Router logic/configuration](#changes-to-router-logic-or-configuration)

The release guidelines for each scenario should generally apply to all release management systems, whether we're deploying to Kubernetes, serverless, or bare metal. Some details might change depending on whether we have a manual deployment process, a continuous delivery process, or something in between. This guide assumes the use of the [Apollo GraphOS][apollographos] for publishing and validating schemas, as well as [Apollo Managed Federation][managedfederation] for propagating supergraph schemas to our Router servers.

[apollographos]: /graphos/

> [managedfederation]: /federation/managed-federation/overview/

Reviewer 2: huh, very neat. never knew you could do this- any reason you don't just use those links in the existing links, though?

Author: I totally can. I don't think I really like this format anyways with just a single link. This was just the way it was from the port. Im guessing this is helpful if you are referencing the same link many times, but in this case that isn't happening

- Making a nullable field non-nullable (this restricts the set of possible output values)
- Removing or changing _unused_ schema elements (no existing operation will be broken)

> This also includes changes to schema metadata like adding or changing a deprecations or descriptions, but we won't consider these changes because they don’t affect runtime behavior.

### Release steps

The release steps for backward compatible changes are as follows:

1. In the subgraph code repository, merge a changeset containing both the schema change and the corresponding resolvers to the release branch, triggering the release pipeline.
2. The release pipeline performs prerelease steps like running tests and schema checks.
3. The release pipeline builds the deployable artifacts:
   - The container, jar, or zip of the runtime code
   - Deployment manifests such as Kubernetes YAML or SAM templates
   - The subgraph SDL document
4. The release pipeline then triggers two "deployments" in sequence.
   1. A rolling deploy of the runtime code.

>    2. A publish of the subgraph SDL document to Apollo GraphOS, which in turn updates the Router with a new supergraph schema (if the subgraph composes successfully; more on that later).

Reviewer 2: Could we make this steps 4 and 5 instead since they run in sequence?

Reviewer: iirc i grouped them to i could refer to both tasks as "Step 4" in the next section

Author: I like it more the way it is than having it laid out as step 4 and 5 for the reason Lenny is pointing out. This is sort of a main point on the whole article, talking about these two steps in particular. We could probably make it stand out more to make this point, but having them under the same step makes sense to me.

5. The release pipeline performs postrelease steps such as smoke tests.

### Sequencing schema publishing with service deployments

During Step 4, our system will most likely end up in an inconsistent state, where the Router's version of the schema is out of sync with the subgraph's version.

When the subgraphs update first, a client _could_ request the new schema elements, and the Router will raise an operation validation error because it is not yet aware of changes.

In practice, however, this rarely matters. Due to the declarative nature of GraphQL, clients must update their operations to use the new schema elements. **Clients aren’t polling our schema's introspection response to discover new fields and immediately start requesting them.** (Also, Apollo recommends [disabling introspection in production][disableprod]!) Only after the system reaches a consistent state will we announce the availability of these new schema elements to clients.

> [disableprod]: https://www.apollographql.com/blog/graphql/security/why-you-should-disable-graphql-introspection-in-production/

Reviewer 2: Same note as above.

[notifications]: /graphos/notifications/schema-change-integration/

### Concurrent subgraph schema changes

Most of the time we can focus on a single subgraph, but it is possible to encounter conflicts when two or more subgraphs change. Consider this sequence of events:

1. Team A makes a change to Subgraph A. The build check succeeds in CI and they merge their code.
2. Team B makes a change to Subgraph B. Their build check also succeeds in CI, even though it actually conflicts with the change in Subgraph A, because the build check runs against _released_ schema definitions.
3. Team B releases their change to production.
4. Team A attempts to release their change, but their new subgraph schema fails to compose and the Router does not update.

The solution is to encounter the build error much earlier in the process. As soon as Team A merges a changeset, Team B should run build checks against that change.

We can accomplish this by creating a special variant of our graph that tracks the `main` branch of each code repository. Apollo GraphOS provides a default variant called `current` for just this use-case. This requires that:

> - Each repository immediately publishes the subgraph schema to the `current` variant each time we merge a change to the `main` branch, even if we won't release the change for a while.

Reviewer 2: Do we want to specifically call out current, or identify using any variant so long as it's consistent across all usages? Basically having a "default" variant.

Reviewer 2: Also - this assumes a monorepo to some extent. CI/CD may differ across teams (e.g. multiple GH repos where main could still be up to date, thus the prior check works, but would fail upon merge). 

Not sure there's a great answer here beyond "use webhooks to rerun checks, however.

> The longer the time between merging and releasing our change, the more important it is to track changes using the `current` variant. On the other hand, if we have a continuous delivery release process and the time between merge and release is relatively short, it's less likely for teams to introduce incompatible changes with unreleased subgraph schemas.

### Rollbacks

Once clients start using the new schema changes, rolling back a subgraph schema change is the same as making a backward incompatible schema change. This may be difficult or impossible to do depending on client usage. For this reason, it's important to test our new schema changes thoroughly before clients start using the new schema in their operations.

If clients are not yet using the schema changes, we most likely don't need to roll them back — they have no effect until operations start using them.

If circumstances require a rollback, the safest process is to revert the change in the subgraph code repository and perform the full release process, publishing the subgraph schema and rolling out the new code together.

### Staged schema change releases

> Instead of attempting rollbacks of schema changes, we may want to release our changes in stages, such as alpha, beta, and generally-available (GA). The "Contracts" feature of GraphOS makes this process simple:

Reviewer 2: Link to Contracts might be nice: https://www.apollographql.com/docs/graphos/delivery/contracts/

> 1. "Tag" new elements of our schema with labels for the targeted release:

Reviewer 2: ```suggestion
1. Mark elements of our schema with the `@tag` directive for the targeted release:
```

   ```graphql
   type Product {
     id: ID!
     newField: String @tag(name: "alpha")
   }
   ```

> 2. Configure our GA variant to be a "contract" that filters out all fields tagged "alpha" or "beta".

Reviewer 2: ```suggestion
2. Configure our GA variant to be a contract that filters out all fields tagged "alpha" or "beta".
```

[clientids]: /graphos/metrics/client-awareness

### Sequencing schema publishing with service deployments

As with a backward compatible release, our system will most likely end up in an inconsistent state, where the Router's version of the schema is out of sync with the subgraph's version.

When the supergraph updates in the Router, it will no longer validate requests that include removed fields, or respond with null values that are now marked non-null.

But with the pre-work outlined above, we've either minimized or completely removed requests that cause these errors.

### Rollbacks

Assuming we've performed the pre-work and reduced usage of relevant types and fields before changing/removing them, rolling back our changes is effectively a backward compatible schema change.

If circumstances require a rollback, the safest process is to revert the change in the subgraph code repository and perform the full release process, publishing the subgraph schema and rolling out the new code together.

> ## Changes to query planning (field migrations, entity ownership changes, etc.)

Reviewer: feels like we should just remove this in favor of the docs: https://www.apollographql.com/docs/federation/entities-advanced

Apollo strongly recommends that we isolate query planning changes from other kinds of schema or implementation changes. All services should be already have working resolvers to support the new query plans.

Consider the scenario of moving a field on an entity from one subgraph (Subgraph A) to another (Subgraph B). After the migration, Subgraph B takes on the responsibility for resolving the field.

```graphql title="Before"
# Subgraph A
type User @key(fields: "id") {
  id: ID!
  fieldToMigrate: String
}

# Subgraph B
type User @key(fields: "id") {
  id: ID! 
}
```

```graphql title="After"
# Subgraph A
type User @key(fields: "id") {
  id: ID!
  fieldToMigrate: String
}

# Subgraph B
type User @key(fields: "id") {
  id: ID!
  fieldToMigrate: String @override(from: "A")
}
```

> For more information on this topic, check out our federation docs on [Migrating entities and fields](/federation/entities-advanced/#migrating-entities-and-fields)

> ### Release steps

Reviewer: Yeah this is all out of date, it should talk about sharable/override. 

I'd be willing to trim this down to just the first two sections — the most important point i was trying to make was that the eventual consistency between the subgraphs and router isn't a significant concern.

Reviewer: and the last three sections aren't really that interesting. they could move to another technote eventually but i think we can just remove them.

Reviewer 2: Agree with Lenny here.

Author: I was leaning the same direction here. Every time i've shared this article, i've always linked to the release steps sections for backwards compatible changes. Ok, I will remove those last 3 sections

When the subgraphs update first, a client _could_ request the new schema elements, and the Router will raise an operation validation error because it is not yet aware of changes.

> In practice, however, this rarely matters. Due to the declarative nature of GraphQL, clients must update their operations to use the new schema elements. **Clients aren’t polling our schema's introspection response to discover new fields and immediately start requesting them.** (Also, Apollo recommends [disabling introspection in production](https://www.apollographql.com/blog/graphql/security/why-you-should-disable-graphql-introspection-in-production/) Only after the system reaches a consistent state will we announce the availability of these new schema elements to clients.

Reviewer 2: ```suggestion
In practice, however, this rarely matters. Due to the declarative nature of GraphQL, clients must update their operations to use the new schema elements. **Clients aren’t polling our schema's introspection response to discover new fields and immediately start requesting them.** (Also, Apollo recommends [disabling introspection in production](https://www.apollographql.com/blog/graphql/security/why-you-should-disable-graphql-introspection-in-production/). Only after the system reaches a consistent state will we announce the availability of these new schema elements to clients.
```
